### PR TITLE
documentation: boardfarm tests integration design review

### PR DESCRIPTION
**Pull request by abelog**
## Description

According to #1116 , every new feature needs a design document.
Migration from `test_flows.py` to Boardfarm for HW testing was discussed and agreed on before #1116. But this activity has to be documented in any case, to avoid effort which we had to apply for documenting `test_flows.py`

This PR proposes documentation which describes the end state of boardfarm integration.
Idea is to modify it while development is going on if something is changed.

Related to #1074 .

### CODE REVIEW

**Comment by abelog** [comment]
Small **note**:
Seems, that images are not rendered properly, as Github looks for them in master branch. But those images are introduced by this PR.

**Review by MuteSpirit — APPROVED** [review]

File: documentation/prplMesh-test-boardfarm.md
Comment (arnout) [inline documentation/prplMesh-test-boardfarm.md]:
```suggestion
Also, some additional abstraction of easyMesh specific entities may ease future work (like CAPI, IEEE1905.1 message).
```

File: documentation/prplMesh-test-boardfarm.md
Comment (arnout) [inline documentation/prplMesh-test-boardfarm.md]: It would be better to use the actual names that we end up using:
```suggestion
	1. UCCSocket
	1. ALEntity
        1. Radio
            1. RadioDocker
            1. RadioHostapd
        1. VirtualAP
            1. VirtualAPDocker
            1. VirtualAPHostapd
```

Bonus points for updating the UML files with these names as well :-)

File: documentation/prplMesh-test-boardfarm.md
Comment (arnout) [inline documentation/prplMesh-test-boardfarm.md]: Actually, I don't think this is true. The classes we add are only useful for prplmesh-specific tests, and the devices we add are only useful to make the link.

File: documentation/prplMesh-test-boardfarm.md
Comment (arnout) [inline documentation/prplMesh-test-boardfarm.md]: I don't think this is relevant in an architecture document that sits in the repository.

Instead, refer to /tests/boardfarm_plugins/README.md

**Review by arnout — APPROVED** [review]
Small things only.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Then i will remove that mention. Having at least a link to those files may help other who seeks for examples of new devices.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Ok. Removing it.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Fixed.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: I Have added a new commit with UML update. Let me know if got it right.

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml
Comment (arnout) [inline documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: All of these don't exist at the moment. get_mac() will never exist - mac is accessed directly. status I'm not convinced if we will add it. tx and rxpower we may add in the future. So I propose to remove status and get_mac(). Also, mark the other functions as abstract please.

File: documentation/images/plantuml/boardfarm_prplMesh_classes.puml
Comment (arnout) [inline documentation/images/plantuml/boardfarm_prplMesh_classes.puml]: This is called wait_for_log instead of check_log. Also, it's an abstract function.

For radios, there is no accessor function. The relation with the Radio class is sufficient.

Also the class is ALEntity, not prplmesh_agent.

**Review by arnout — APPROVED** [review]
The last commit is still not entirely correct. Instead of explaining it all, I'm just going to fix it myself. Therefore, I'm approving and reassigning to myself.

**Review by rmelotte — APPROVED** [review]

File: documentation/prplMesh-test-boardfarm.md
Comment (tomereli) [inline documentation/prplMesh-test-boardfarm.md]:
```suggestion
1. Abstraction of test setup
```

File: documentation/prplMesh-test-boardfarm.md
Comment (tomereli) [inline documentation/prplMesh-test-boardfarm.md]: You already use "provides".
```suggestion
1. Basic device abstraction classes
```

File: documentation/prplMesh-test-boardfarm.md
Comment (tomereli) [inline documentation/prplMesh-test-boardfarm.md]: Such as? Please add an example if you're mentioning tests

File: documentation/prplMesh-test-boardfarm.md
Comment (tomereli) [inline documentation/prplMesh-test-boardfarm.md]:
```suggestion
1. DUT connectivity (ssh, serial, telnet, etc.)
```

File: documentation/prplMesh-test-boardfarm.md
Comment (tomereli) [inline documentation/prplMesh-test-boardfarm.md]:
```suggestion
1. Test-suite organization
```
But please, explain what this means. If you need to say a sentence about the YAML file describing a test-suite, than do so.

File: documentation/prplMesh-test-boardfarm.md
Comment (tomereli) [inline documentation/prplMesh-test-boardfarm.md]: This sentence is unclear. I would expect to expand the base classes with EasyMesh device, which can be an agent or controller, so:
```
We expand base classes with the representation of `EasyMesh` devices capable of running as Controller, Agent, or both.
```
What does the sniffer has to do with it?

File: documentation/prplMesh-test-boardfarm.md
Comment (tomereli) [inline documentation/prplMesh-test-boardfarm.md]: Better way to say this IMO:
Note - normally, it is expected that the prplmehs IPK is deployed to the DUT (Device Under Test), or that the DUT is already flashed with prplmesh embedded in the image.
For this reason, LAN connection for flashing in uboot is not required and is deliberately missing from the scheme.

**Review by tomereli — APPROVED** [review]
Mainly comments about the README.md file, but nothing critical so approving.

**Comment by tomereli** [comment]
Approved but added the don't merge tag in case @abelog you want to fix my comments.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Those are generic tests, like: check memory usage, check CPU usage and etc. They are not much of use to us. So i will remove that mention.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: It  allows us to have logical grouping of tests and bulk execution of them. As result, all reports and logs are formed for whole test suite (still, logs are also available per test so you don't have to look general log for specific failed test). We can also execute each test individually, but it's not convenient.
It's not is YML, it's just a plain text file with some formatting. [Example of such `.cfg` file in boardfarm repository](https://github.com/mattsm/boardfarm/blob/master/boardfarm/testsuites.cfg).

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Applied.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Applied.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Fixed!

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Sniffer is not EasyMesh entity, it's standalone docker image/device in setup. Order of words has confused.
Moved Sniffer mention before the mention of EasyMesh-specific entities, to avoid confusion that Sniffer has anything to do with EasyMesh specific entities.

File: documentation/prplMesh-test-boardfarm.md
Comment (abelog) [reply documentation/prplMesh-test-boardfarm.md]: Taken in with some fixes of mine.

**Comment by abelog** [comment]
@tomereli, i have applied suggestions of yours( see 7e4e7e1c698b1ef09c5fc09feb3966b75546762e). Please approve again if it's fine by you. Only after that i will rebase and merge  this document.

**Comment by tomereli** [comment]
> @tomereli, i have applied suggestions of yours( see 7e4e7e1c698b1ef09c5fc09feb3966b75546762e). Please approve again if it's fine by you. Only after that i will rebase and merge  this document.

I already approved, you can merge

**Comment by rmelotte** [comment]
@mergifyio rebase

**Pull request merged**
